src/platform_impl/web/web_sys/event.rs

```diff
@@ -61,190 +62,71 @@ pub fn mouse_scroll_delta(event: &WheelEvent) -> Option<MouseScrollDelta> {
-        "NumpadComma" => VirtualKeyCode::NumpadComma,
-        "NumpadEnter" => VirtualKeyCode::NumpadEnter,
-        "NumpadEquals" => VirtualKeyCode::NumpadEquals,
-        "OEM102" => VirtualKeyCode::OEM102,
-        "Period" => VirtualKeyCode::Period,
-        "PlayPause" => VirtualKeyCode::PlayPause,
-        "Power" => VirtualKeyCode::Power,
-        "PrevTrack" => VirtualKeyCode::PrevTrack,
-        "AltRight" => VirtualKeyCode::RAlt,
-        "BracketRight" => VirtualKeyCode::RBracket,
-        "ControlRight" => VirtualKeyCode::RControl,
-        "ShiftRight" => VirtualKeyCode::RShift,
-        "MetaRight" => VirtualKeyCode::RWin,
-        "Slash" => VirtualKeyCode::Slash,
-        "Sleep" => VirtualKeyCode::Sleep,
-        "Stop" => VirtualKeyCode::Stop,
-        "NumpadSubtract" => VirtualKeyCode::NumpadSubtract,
-        "Sysrq" => VirtualKeyCode::Sysrq,
-        "Tab" => VirtualKeyCode::Tab,
-        "Underline" => VirtualKeyCode::Underline,
-        "Unlabeled" => VirtualKeyCode::Unlabeled,
-        "AudioVolumeDown" => VirtualKeyCode::VolumeDown,
-        "AudioVolumeUp" => VirtualKeyCode::VolumeUp,
-        "Wake" => VirtualKeyCode::Wake,
-        "WebBack" => VirtualKeyCode::WebBack,
-        "WebFavorites" => VirtualKeyCode::WebFavorites,
-        "WebForward" => VirtualKeyCode::WebForward,
-        "WebHome" => VirtualKeyCode::WebHome,
-        "WebRefresh" => VirtualKeyCode::WebRefresh,
-        "WebSearch" => VirtualKeyCode::WebSearch,
-        "WebStop" => VirtualKeyCode::WebStop,
-        "Yen" => VirtualKeyCode::Yen,
-        _ => return None,
-    })
+fn cached_string<S: AsRef<str>>(string: S) -> &'static str {
+    use std::collections::HashSet;
+    use std::sync::Mutex;
+
+    use once_cell::sync::Lazy;
+
+    static STRING_CACHE: Lazy<Mutex<HashSet<&'static str>>> =
+        Lazy::new(|| Mutex::new(HashSet::new()));
+
+    let string = string.as_ref();
+    let mut cache = STRING_CACHE.lock().unwrap();
+    if let Some(string) = cache.get(string) {
+        string
+    } else {
+        // borrowck couldn't quite figure out this one on its own
+        let string: &'static str = Box::leak(String::from(string).into_boxed_str());
+        cache.insert(string);
+        string
+    }
 }
 
-pub fn keyboard_modifiers(event: &KeyboardEvent) -> ModifiersState {
-    let mut m = ModifiersState::empty();
-    m.set(ModifiersState::SHIFT, event.shift_key());
-    m.set(ModifiersState::CTRL, event.ctrl_key());
-    m.set(ModifiersState::ALT, event.alt_key());
-    m.set(ModifiersState::LOGO, event.meta_key());
-    m
+pub fn key_location(event: &KeyboardEvent) -> KeyLocation {
+    match event.location() {
+        KeyboardEvent::DOM_KEY_LOCATION_LEFT => KeyLocation::Left,
+        KeyboardEvent::DOM_KEY_LOCATION_RIGHT => KeyLocation::Right,
+        KeyboardEvent::DOM_KEY_LOCATION_NUMPAD => KeyLocation::Numpad,
+        KeyboardEvent::DOM_KEY_LOCATION_STANDARD => KeyLocation::Standard,
+        _ => KeyLocation::Standard,
```

Comment:
Continuation of https://github.com/rust-windowing/winit/pull/2662#discussion_r1186274430.
> We don't have unknown, and I don't think we should have it.

This will only be triggered if the specification changes and a new location is introduced, so it's kinda wrong what we are returning here, but I agree.

> We might have `unreachable!()`, but I'm afraid to do so for web, so let's keep the `Standard`, it should be more or less safe.

`unreachable!()` would work right now, but the Web changes a lot.

---

We could at least print an error or warning message with `log`? Does that sound appropriate?

Reply:
I think that should be fine. We might add `Unknown` for sure.

Reply:
I don't think we can reasonably add `Unknown`. The issue is that even the specification kinda defines `Standard` as "all the rest"
> DOM_KEY_LOCATION_STANDARD
>    The key activation MUST NOT be distinguished as the left or right version of the key, and (other than the NumLock key) did not originate from the numeric keypad (or did not originate with a virtual key corresponding to the numeric keypad).

I don't think we could distinguish between `Standard` and `Unknown` on the desktop implementations (Win/X11/Wayland/macOS).

Reply:
According to the specification there isn't any other key location, so `unreachable!()` is the *correct* thing to do here. 

But if a new key location is added later, which is not out of the question, it would fail, but assuming `DOM_KEY_LOCATION_STANDARD` in that case is definitely **wrong**. So `KeyboardEvent::Unknown` makes sense here.

I'm currently in favor of **not** adding a `KeyboardEvent::Unknown` and instead just returning `KeyLocation::Standard` and use `log::error!()`. There is no need to burden the other backends only because *it's possible* that another key location is added.
> I don't think we could distinguish between `Standard` and `Unknown` on the desktop implementations (Win/X11/Wayland/macOS).

Another nail in the coffin of `KeyboardEvent::Unknown`.

That said it 's unlikely that a new key location will be added, the only one that was being considered is a gamepad, which was agreed to be moved to a dedicated game pad API: https://github.com/w3c/uievents/issues/79. Which has already happened: https://caniuse.com/gamepad.

---

**Conclusion**
We don't need to worry about this.